shell/platform/embedder/embedder.h

```diff
@@ -611,6 +611,65 @@ typedef struct {
   int64_t buttons;
 } FlutterPointerEvent;
 
+typedef enum {
+  kFlutterKeyEventKindUp = 1,
+  kFlutterKeyEventKindDown,
+  kFlutterKeyEventKindRepeat,
+} FlutterKeyEventKind;
+
+/// A structure to represent a change of state of a key.
+///
+/// Sending `FlutterKeyEvent` via `FlutterEngineSendKeyEvent` results in a
+/// corresponding `FlutterKeyEvent` to be dispatched in the framework. It is
+/// embedder's responsibility to ensure the regularity of sent events, since the
+/// framework only performs simple one-to-one mapping. The events must conform
+/// the following rules:
+///
+///  * Each key press sequence shall consist of one key down event (`kind` being
+///    `kFlutterKeyEventKindDown`), zero or more repeat events, and one key up
+///    event, representing a physical key button being pressed, held, and released.
+///  * All events throughout a key press sequence shall have the same `physical`
+///    and `logical`. Having different `character`s is allowed.
+typedef struct {
+  /// The size of this struct. Must be sizeof(FlutterKeyEvent).
+  size_t struct_size;
+  /// The timestamp at which the pointer event was generated. The timestamp
+  /// should be specified in microseconds and the clock should be the same as
+  /// that used by `FlutterEngineGetCurrentTime`.
+  double timestamp;
+  /// The event kind.
+  FlutterKeyEventKind kind;
+  /// The USB HID code for the physical key of the event.
+  ///
+  /// For the full definition and list of pre-defined physical keys, see
+  /// `PhysicalKeyboardKey` from the framework.
+  uint64_t physical;
+  /// The key ID for the logical key of this event.
+  ///
+  /// For the full definition and a list of pre-defined logical keys, see
+  /// `LogicalKeyboardKey` from the framework.
+  uint64_t logical;
+  /// Null-terminated character input from the event. Can be null. Ignored for
+  /// up events.
+  const char* character;
+  /// True if this event does not correspond to a native event.
+  ///
+  /// The embedder is likely to skip events and/or construct new events that do
+  /// not correspond to any native events in order to conform the regularity
+  /// of events (as documented in `FlutterKeyEvent`). An example is when a key up
+  /// is missed due to loss of window focus, on a platform that provides query to
+  /// key pressing status, the embedder might realize that the key has been
+  /// released at the next key event, and should construct a synthesized up event
+  /// immediately before the actual event.
+  ///
+  /// An event being synthesized means that the framework will not trust the
+  /// `timestamp` of the event.
```

Comment:
I'm not sure what this means to a client of this function. It suggests that maybe `timestamp` doesn't need to be set for synthesized events, but I don't think that's actually true. Maybe instead say something like "A synthesized event should use the timestamp of the time at which it was created, since the missed event's actual timestamp is unknown."